File: Tellma/Entities/DashboardDefinition.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System;

namespace Tellma.Entities
{
    [StrongEntity]
    [EntityDisplay(Singular = "DashboardDefinition", Plural = "DashboardDefinitions")]
    public class DashboardDefinitionForSave<TWidget, TRole> : EntityWithKey<int>
    {
        [Display(Name = "Code")]
        [NotNull]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Code { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Primary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Secondary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title2 { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Ternary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title3 { get; set; }

        [Display(Name = "DashboardDefinition_AutoRefreshPeriodInMinutes")]
        [NotNull]
        [AlwaysAccessible]
        public int? AutoRefreshPeriodInMinutes { get; set; }

        [Display(Name = "Definition_ShowInMainMenu")]
        [AlwaysAccessible]
        public bool? ShowInMainMenu { get; set; }

        [AlwaysAccessible]
        public string MainMenuSection { get; set; }

        [Display(Name = "MainMenuIcon")]
        [AlwaysAccessible]
        public string MainMenuIcon { get; set; }

        [Display(Name = "MainMenuSortKey")]
        [AlwaysAccessible]
        public decimal? MainMenuSortKey { get; set; }

        [Display(Name = "DashboardDefinition_Widgets")]
        [ForeignKey(nameof(DashboardDefinitionWidget.DashboardDefinitionId))]
        [AlwaysAccessible]
        public List<TWidget> Widgets { get; set; }

        [Display(Name = "Definition_Roles")]
        [ForeignKey(nameof(DashboardDefinitionRole.DashboardDefinitionId))]
        [AlwaysAccessible]
        public List<TRole> Roles { get; set; }
    }

    public class DashboardDefinitionForSave : DashboardDefinitionForSave<DashboardDefinitionWidgetForSave, DashboardDefinitionRoleForSave>
    {

    }

    public class DashboardDefinition : DashboardDefinitionForSave<DashboardDefinitionWidget, DashboardDefinitionRole>
    {
        [Display(Name = "CreatedAt")]
        [NotNull]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        [NotNull]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        [NotNull]
        public DateTimeOffset? ModifiedAt { get; set; }

        [Display(Name = "ModifiedBy")]
        [NotNull]
        public int? ModifiedById { get; set; }

        // For Query

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(ModifiedById))]
        public User ModifiedBy { get; set; }
    }

    public class DashboardDefinitionWidgetForSave : EntityWithKey<int>
    {
        [Display(Name = "DashboardDefinition_ReportDefinition")]
        [NotNull]
        public int? ReportDefinitionId { get; set; }

        [Display(Name = "DashboardDefinition_OffsetX")]
        [NotNull]
        public int? OffsetX { get; set; }

        [Display(Name = "DashboardDefinition_OffsetY")]
        [NotNull]
        public int? OffsetY { get; set; }

        [Display(Name = "DashboardDefinition_Width")]
        [NotNull]
        public int? Width { get; set; }

        [Display(Name = "DashboardDefinition_Height")]
        [NotNull]
        public int? Height { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Primary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Secondary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title2 { get; set; }

        [MultilingualDisplay(Name = "Title", Language = Language.Ternary)]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Title3 { get; set; }

        [Display(Name = "DashboardDefinition_AutoRefreshPeriodInMinutes")]
        [NotNull]
        [AlwaysAccessible]
        public int? AutoRefreshPeriodInMinutes { get; set; }
    }

    public class DashboardDefinitionWidget : DashboardDefinitionWidgetForSave
    {
        [NotNull]
        public int? DashboardDefinitionId { get; set; }

        [NotNull]
        public int? Index { get; set; }

        [Display(Name = "DashboardDefinition_ReportDefinition")]
        [ForeignKey(nameof(ReportDefinitionId))]
        public ReportDefinition ReportDefinition { get; set; }
    }


    public class DashboardDefinitionRoleForSave : EntityWithKey<int>
    {
        [Display(Name = "Definition_Role")]
        [NotNull]
        public int? RoleId { get; set; }
    }

    public class DashboardDefinitionRole : DashboardDefinitionRoleForSave
    {
        [NotNull]
        public int? DashboardDefinitionId { get; set; }

        [Display(Name = "Definition_Role")]
        [ForeignKey(nameof(RoleId))]
        public Role Role { get; set; }
    }
}

File: Tellma.Api.Web/ClientApp/src/app/data/entities/dashboard-definition.ts
// tslint:disable:variable-name
// tslint:disable:max-line-length
import { EntityForSave } from './base/entity-for-save';
import { WorkspaceService } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { SettingsForClient } from '../dto/settings-for-client';
import { EntityDescriptor } from './base/metadata';
import { EntityWithKey } from './base/entity-with-key';
import { MainMenuSection, MainMenuIcon, mainMenuSectionPropDescriptor, mainMenuIconPropDescriptor, mainMenuSortKeyPropDescriptor } from './base/definition-common';
import { TimeGranularity } from './base/metadata-types';

export interface DashboardDefinitionForSave<
    TWidget = DashboardDefinitionWidgetForSave,
    TRole = DashboardDefinitionRoleForSave> extends EntityForSave {

    Code?: string;
    Title?: string;
    Title2?: string;
    Title3?: string;
    AutoRefreshPeriodInMinutes?: number;
    MainMenuSection?: MainMenuSection;
    MainMenuIcon?: MainMenuIcon;
    MainMenuSortKey?: number;

    Widgets?: TWidget[];
    Roles?: TRole[];
}

export interface DashboardDefinition extends DashboardDefinitionForSave<
        DashboardDefinitionWidget, DashboardDefinitionRole> {
    ShowInMainMenu?: boolean;
    CreatedAt?: string;
    CreatedById?: number | string;
    ModifiedAt?: string;
    ModifiedById?: number | string;
}

export interface DashboardDefinitionWidgetForSave extends EntityForSave {
    ReportDefinitionId?: number;
    OffsetX?: number;
    OffsetY?: number;
    Width?: number;
    Height?: number;
    Title?: string;
    Title2?: string;
    Title3?: string;
    AutoRefreshPeriodInMinutes?: number;
}

// tslint:disable-next-line:no-empty-interface
export interface DashboardDefinitionWidget extends DashboardDefinitionWidgetForSave {
    DashboardDefinitionId?: number;
}

export interface DashboardDefinitionRoleForSave extends EntityForSave {
    RoleId?: number;
}

// tslint:disable-next-line:no-empty-interface
export interface DashboardDefinitionRole extends DashboardDefinitionRoleForSave {
    DashboardDefinitionId?: number;
}


const _select = ['', '2', '3'].map(pf => 'Title' + pf);
let _settings: SettingsForClient;
let _cache: EntityDescriptor;

export function metadata_DashboardDefinition(wss: WorkspaceService, trx: TranslateService): EntityDescriptor {
    const ws = wss.currentTenant;
    // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
    if (ws.settings !== _settings) {
        _settings = ws.settings;
        _cache = {
            collection: 'DashboardDefinition',
            titleSingular: () => trx.instant('DashboardDefinition'),
            titlePlural: () => trx.instant('DashboardDefinitions'),
            apiEndpoint: 'dashboard-definitions',
            masterScreenUrl: 'dashboard-definitions',
            select: _select,
            orderby: () => ws.isSecondaryLanguage ? [_select[1], _select[0]] : ws.isTernaryLanguage ? [_select[2], _select[0]] : [_select[0]],
            inactiveFilter: null,
            format: (item: EntityWithKey) => (ws.getMultilingualValueImmediate(item, _select[0]) || trx.instant('Untitled')),
            formatFromVals: (vals: any[]) => (ws.localize(vals[0], vals[1], vals[2]) || trx.instant('Untitled')),
            properties: {
                Id: { noSeparator: true, datatype: 'numeric', control: 'number', label: () => trx.instant('Id'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
                Code: { datatype: 'string', control: 'text', label: () => trx.instant('Code') },
                Title: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.primaryPostfix },
                Title2: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.secondaryPostfix },
                Title3: { datatype: 'string', control: 'text', label: () => trx.instant('Title') + ws.ternaryPostfix },
                AutoRefreshPeriodInMinutes: { datatype: 'numeric', control: 'number', label: () => trx.instant('DashboardDefinition_AutoRefreshPeriodInMinutes'), minDecimalPlaces: 0, maxDecimalPlaces: 0, noSeparator: false },
                MainMenuSection: mainMenuSectionPropDescriptor(trx),
                MainMenuIcon: mainMenuIconPropDescriptor(trx),
                MainMenuSortKey: mainMenuSortKeyPropDescriptor(trx),
                ShowInMainMenu: { datatype: 'bit', control: 'check', label: () => trx.instant('Definition_ShowInMainMenu') },
                CreatedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('CreatedAt'), granularity: TimeGranularity.minutes },
                CreatedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('CreatedBy'), foreignKeyName: 'CreatedById' },
                ModifiedAt: { datatype: 'datetimeoffset', control: 'datetime', label: () => trx.instant('ModifiedAt'), granularity: TimeGranularity.minutes },
                ModifiedBy: { datatype: 'entity', control: 'User', label: () => trx.instant('ModifiedBy'), foreignKeyName: 'ModifiedById' }
            }
        };

        if (!ws.settings.SecondaryLanguageId) {
            delete _cache.properties.Title2;
        }

        if (!ws.settings.TernaryLanguageId) {
            delete _cache.properties.Title3;
        }
    }

    return _cache;
}
